### ISSUE TITLE
ServiceStack in Xamarin

### ISSUE THREAD

**Issue by Ingenator** [ServiceStack/Issues#451]
I'm having a werid issue with the recent updates. I'm Using Visual Studio Enterprise 2015 and Xamarin.Forms 2.3

I'm using Service Stack, and everything works cool but here's when the story goes werid.

I have a REST API with service stack, and the app is the client, but when I request a method, no matter if its using POST or GET the API response is correct but when it enters the app it goes to empty with a custom property type. I know it is correct because logs on the API says so and I know it reaches the app using WireShark and the values are there, furthermore it was working before I upgraded from Visual Studio 2013 to Visual Studio 2015. I formatted my PC and resintalled everything, still doesn't work.

If I run the same code (client methods) in a windows forms it works as expected but running it on Xamarin.Forms it does not.

The issue happens with a custom type I created but when I use the generic type

The API has a DataProperty property I use to send some encoded data.

```
public class DataProperty
{
    protected string base64Value;

    public DataProperty();

    public string Value { get; }

    public byte[] Get();
    public DataProperty Set(byte[] data);

    public static implicit operator DataProperty(byte[] data);
    public static implicit operator byte[] (DataProperty PropertyValue);
}
```

```
public class DataProperty<T> : DataProperty
{
    public DataProperty();

    public DataProperty<T> Set(byte[] data);

    public static implicit operator DataProperty<T>(byte[] data);
    public static implicit operator byte[] (DataProperty<T> PropertyValue);
}
```
Service stack json returns:

```
[0:] {
  "Token": {
    "Value": "WFE1Au5DoIb0sEttlD3h08Xg+E+KwrZ5tFusJqP1TsSY6D6PACjcBAiub+ydX84Wa95Gu318ziDXssUnwe0fWE0L/0A6h5IcLEYk8xGxKdW3ooehjrOR/7KwEypZRcmVI6oKX75id21npQDJZP8Y4U18gBeJuuKjpckeL8b54U8="
  },
  "Id": {
    "Value": ""
  },
  "AId": {
    "Value": ""
  },
  "Iden": {
    "Value": ""
  },
  "ACID": 1522,
  "ResponseStatus": null
}
```
For the record, the method response properties are the following:

```
    public class MyMethodResponse 
    {
        public DataProperty Token { get; set; }

        public DataProperty<long> Id { get; set; }

        public DataProperty<double> AId { get; set; }

        public DataProperty<string> Iden { get; set; }

        public long ACID { get; set; }
    }
```

As I said, if I run exactly the same code on a PCL+Windows Forms(or even console) app it works correctly but running it on PCL+Xamarin.Forms it doesn't (using emulator or phisical phone) Tried XF 2.1,2.2, 2.3.

The client implementation method is an async Task<bool> and its like the following:

```
        MyMethodRequest request = new MyMethodRequest();
        using (JsonServiceClient client = new JsonServiceClient(serviceURL))
        {
            try
            {
                var result = await client.PostAsync(request);
                Debug.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));
                return true;
            }
            catch (Exception e)
            {
                Debug.WriteLine("Error:[" + e.Message + "]");
                return false;
            }
        }
```

My machine has Windows 10 Pro, Dev envirorment is Visual Studio 2015
The Server is a Windows Server 2008 R2 

Regardless if its on my machine or the server the problem persists.

Any ideas?

**Comment by mythz** [comment]
Can you provide a small, stand-alone example that reproduces the issue, you likely wont need to call a Service, can you provide an example of a DTO that fails to serialize in Xamarin.Forms?

Also why are you wrapping values in `DataProperty<T>`? Your DTO/POCO's shouldn't have behaviors.

**Comment by Ingenator** [comment]
Thanks mythz!

I'm Working on the stand-alone example, I have the client, PCL+WinForms and the same PCL+Xamari.Forms but I'm working on the example service.

I'm wrapping values into a `DataProperty<T>` beacuase I need to encrypt/decrypt it so I know the Type of the encrypted Data to work correctly with it.

**Comment by mythz** [comment]
Seems strange to encrypt fields individually instead of just encrypting the whole DTO.

BTW I've just included [Encrypted Messaging](https://github.com/ServiceStack/ServiceStack/wiki/Encrypted-Messaging) support in the Xamarin iOS/Android and Mac client platform builds so you could now use that on iOS/Android if needed.

It's available from v4.0.61 that's now [available on MyGet](https://github.com/ServiceStack/ServiceStack/wiki/MyGet).

**Comment by Ingenator** [comment]
Its complicated, is a bank-like app so we manage AES and RSA encryptions with dynamic keys per device and I can't have an static key to encrypt/decrypt data.

Here is the example of the backend
https://github.com/Ingenator/ServiceStackBug

**Comment by Ingenator** [comment]
Finally, here is the client https://github.com/Ingenator/ServiceStackBugApp

You can run it on the Droid client or the WFExample

1.- You need to bring up the service in here https://github.com/Ingenator/ServiceStackBug
2.- Bring up the clients, the examples contain the Droid and the Windows form clients, executing the client on the Droid (Xamarin.Forms) it does execute but no values are returned on the other hand executing it on hte DesktopApp (Windows Forms) it does exectue and shows the values.

**Comment by xplicit** [comment]
I ran ExampleApp.Droid application and got the following output. Are these values correct? Visual Studio Version shows info:

    Microsoft Visual Studio Community 2015
    Version 14.0.25123.00 Update 2

    Xamarin   4.0.3.214 (0dd817c)
    Xamarin.Android   6.0.3.5 (a94a03b)
    Xamarin.iOS   9.6.1.8 (3a25bf1)

Output:

```
06-29 23:29:17.198 I/mono-stdout(  802): {
  "FistValueResponse": {
06-29 23:29:17.207 I/mono-stdout(  802):   "FistValueResponse": {
    "Value": "MQAzAA=="
06-29 23:29:17.207 I/mono-stdout(  802):     "Value": "MQAzAA=="
  },
06-29 23:29:17.220 I/mono-stdout(  802):   },
  "SecondValueReponse": {
06-29 23:29:17.229 I/mono-stdout(  802):   "SecondValueReponse": {
    "Value": "MQAzAA=="
  },
  "ThirdValueResponse": {
    "Value": "MQAzAA=="
06-29 23:29:17.237 I/mono-stdout(  802):     "Value": "MQAzAA=="
06-29 23:29:17.237 I/mono-stdout(  802):   },
06-29 23:29:17.237 I/mono-stdout(  802):   "ThirdValueResponse": {
06-29 23:29:17.237 I/mono-stdout(  802):     "Value": "MQAzAA=="
  },
06-29 23:29:17.248 I/mono-stdout(  802):   },
  "PlainValueResponse": 13,
[0:] {
  "FistValueResponse": {
    "Value": "MQAzAA=="
  },
  "SecondValueReponse": {
    "Value": "MQAzAA=="
  },
  "ThirdValueResponse": {
    "Value": "MQAzAA=="
  },
  "PlainValueResponse": 13,
  "ResponseStatus": null
}
06-29 23:29:17.259 I/mono-stdout(  802):   "PlainValueResponse": 13,
06-29 23:29:17.259 I/mono-stdout(  802):   "ResponseStatus": null
06-29 23:29:17.259 I/mono-stdout(  802): }
  "ResponseStatus": null
}


```

**Comment by Ingenator** [comment]
Yeah does values are right,

I have the same visual studio version but on the Xamarin side I have different ones:

Xamarin 4.1.1.3 (34a92cd)
Xamarin.Android 6.1.1.1 (7db2aac)
Xamarion.iOS 9.8.1.4 (3cf8aae)

**Comment by Ingenator** [comment]
Ok I downgraded the versions of Xamarin and it worked correctly but using the current stable versions doesn't.

**Comment by mythz** [comment]
If it works in .NET and in a previous release of Xamarin, it's a regression Xamarin, we'll keep looking to see if we can identify the issue but the fix is going to need to be in Xamarin's libraries.

**Comment by xplicit** [comment]
Xamarin which comes with Visual Studio Update 3 also has this issue.

Xamarin 4.1.0.530 (2e39740)
Xamarin.Android 6.1.0.71 (4e27558)

**Comment by mythz** [comment]
Hi the issue is because Value doesn't have a public setter, i.e:

```
public string Value { get; set; }
```

Where it works when Value has a public setter. DTO's should have properties with public setters/getters in order to retrieve and populate the property during serialization. There is a change from the underyling Mono's reflection API's that causes the difference in behavior, but you shouldn't be relying on serialization of private setters.

**Comment by xplicit** [comment]
Created bug about Xamarin regression in Xamarin Bugzilla

https://bugzilla.xamarin.com/show_bug.cgi?id=42357

**Comment by Ingenator** [comment]
Thank you guys! I was diggin in Xamarin's fourms to report the bug but you did it!

Thats so sad I can't use that private setter in particular, but finally we knew what the issue was.

**Comment by xplicit** [comment]
@Ingenator Support for properties with private setter is added via this commit https://github.com/ServiceStack/ServiceStack.Text/commit/1eb5254f8d98533885e95395b7bb72ca280ba0c1 and is available in v4.0.61 on MyGet

It works with Xamarin v4.1 and now your sample on android produces same results as windows forms application.

**Comment by Ingenator** [comment]
Awesome! thank you guys! keep up the good work!!
